### PR TITLE
M81 - Conditional Delays

### PR DESCRIPTION
**Pull request by zeleps**
In response to FR #23235, I added two switches to M81. 

M81 D[seconds] provides a minimum delay after the execution of M81 before any power off action is taken. 
M81 S1 (same as in RepRapFirmware) will wait until cooldown (all heaters are below auto fan temperature). 

- If both switches are specified, both conditions must be met to power off. 
- The hardcoded 1'' delay after power off initiates (to properly shutdown motors and heaters) is preserved, thus power off will occur approximately 1'' after the specified delay (or after fans are off).
- Any long-running gcode issued after M81, might stall power off until it completes (i.e. G28 will complete homing even if M81 timer elpases).
- Sending M80 (without the S flag) in time, cancels any pending power off.

### Requirements

S1 depends on the auto fan feature (if no auto fan enabled, it's ignored).

### Benefits

Allows users to shut down their external controllers (e.g. Raspberry Pi) gracefully before powering off a shared PSU. Also allows proper cooldown after unattended prints that power off the printer when they end.

### Configurations

Needs PSU_CONTROL.

### Related Issues

FR #23235

### CODE REVIEW

File: Marlin/src/module/temperature.h [hunk 1]
```diff
@@ -993,7 +997,10 @@ class Temperature {
       static int16_t read_max_tc(TERN_(HAS_MULTI_MAX_TC, const uint8_t hindex=0));
     #endif
 
-    static void update_autofans();
```
Comment (zeleps) [inline Marlin/src/module/temperature.h:996]: This was causing a compilation error if HAS_AUTO_FAN was undefined. Now fixed.

**Comment by thinkyhead** [comment]
Marlin already has a feature to wait a period of time before turning off power, as part of the `Power` class when `POWER_OFF_DELAY` and `AUTO_POWER_CONTROL` options are enabled. The `Power::is_power_needed` method checks whether power should be remain on (or should be turned on). That would be the class and the method to extend with the logic being proposed here.

**Comment by zeleps** [comment]
Ok, I can take over and move the timer there if you want, I guess you have better things to do.

**Comment by thinkyhead** [comment]
Note: The capability to wait for cooldown is not dependent on the presence of an auto fan. An always-on fan will also cool the nozzle more quickly if the power is left on.

**Comment by zeleps** [comment]
That's true, a temperature threshold will be needed for that, although a timed power off will more or less cover it.

**Comment by thinkyhead** [comment]
Ok, I'm done messing about for now. I'm sure it's not exactly the same as RRF, but it's a start. We don't have to match the G-code of RRF in all respects, as there is a different paradigm here, but these additions are kind of useful. I would guess that RRF also makes the cooldown temperatures configurable with a G-code, and we don't necessarily need to do that here.

File: Marlin/src/gcode/control/M80_M81.cpp [hunk 2]
```diff
@@ -79,11 +79,19 @@ void GcodeSuite::M81() {
...
-    #endif
+  #if BOTH(HAS_FAN, PROBING_FANS_OFF)
+    thermalManager.fans_paused = false;
+    ZERO(thermalManager.saved_fan_speed);
+  #endif
+
+  LCD_MESSAGE_F(MACHINE_NAME " " STR_OFF ".");
+
+  #if ENABLED(POWER_OFF_TIMER)
+    if (parser.seenval('D')) { powerManager.setPowerOffTimer(SEC_TO_MS(parser.value_ushort())); return; }
```
Comment (zeleps) [inline Marlin/src/gcode/control/M80_M81.cpp]: This does not allow a combination of the two switches, D will always take precedence. You can't cover a scenario where a possible cooldown is wanted, but a minimum delay is necessary for graceful shutdown.

**Comment by stef-ladefense** [comment]
hi,
with the 2.12 and beta up, the M81 not work.
found bug, the bool 'psu_on' is not init : please add in power.cpp, line 46
`bool Power::psu_on = true;`
and now M81 work !
no time for PR, sorry
big kiss

**Comment by thinkyhead** [comment]
That's peculiar. Marlin already does this in `setup()`:
```cpp
  #if ENABLED(PSU_CONTROL)
    SETUP_RUN(powerManager.init());
  #endif
```
…which calls…
```cpp
void Power::init() {
  psu_on = ENABLED(PSU_DEFAULT_OFF);              // Set opposite state to get full power_off/on
  TERN(PSU_DEFAULT_OFF, power_off(), power_on());
}
```
Therefore:
- If you have `PSU_DEFAULT_OFF` enabled it causes `psu_on` to be initialized as `true` and then `power_off()` is called which turns off the power and sets it `false`.
- If you have `PSU_DEFAULT_OFF` disabled it causes `psu_on` to be initialized as `false` and then `power_on()` is called which turns on the power and sets it `true`.

If you don't enable `PSU_CONTROL` at all then various things are turned off by `M81`, but there is no `psu_on` flag and control of the PSU is not available.

**Comment by stef-ladefense** [comment]
@thinkyhead 

yes, but regardless of the value of psu_on that is assigned to it on this line (I could very well have put false), it is because this variable was only declared but not initialized in power.cpp,
and the ways via ENABLED() or whatever did nothing on it.
moreover the C standard indicates cases where an uninitialized variable can behave indefinitely.

**Comment by thinkyhead** [comment]
On these platforms "uninitialized" variables are always zero / false. Meanwhile, as pointed out, the variable is initialized prior to its being used. So, again, there is no reason for its being initialized as "true" to have any impact on behavior, which is why I consider it weird. Capice?

**Pull request merged**